### PR TITLE
Reading known addresses in blockchain-api

### LINKED ISSUE #426: Return known names and pictures from the  blockchain-api
### What would you like?

We have on Firebase a growing node called `addressesExtraInfo`. This is growing by at least one address a week for balance-based rewards distributions and there are other projects that have shown interest in adding their name and picture on Discord. The current solution is not ideal because it requires users to be listening on this node at all times and load all existing known addresses, even if they are not using any. As the number of known addresses grows, this will require more and more bandwidth to load and maintain. It sometimes briefly shows transactions as 'Unknown' if they are loaded faster than the known addresses from Firebase.

The goal of this issue is to return the name and picture from the `blockchain-api` directly.

### PR DESCRIPTION
**Pull request by dievazqu**
### Description

Avoid reading known addresses from firebase in the client, but receiving that info from the blockchain-api

### Tested

Changes were tested running the app locally. But also with unit and e2e tests.

### Related issues

* Fixes #426

### Backwards compatibility

With these changes, we are not showing  anymore known addresses while receiving notifications, since it's not coming from the blockchain-api, that will be done in another PR. Probably the best would be to wait until the other PR is ready to merge to merge both at the same time so there is no unexpected behavior.

Note: If this PR seems big, I have created another [PR](https://github.com/valora-inc/wallet/pull/1110) with most of the changes on the mobile side (but it might be a bit outdated after some changes I did in this PR). However, both PRs should be merged together.

### CODE REVIEW

File: packages/mobile/src/recipients/recipient.ts [hunk 1]
```diff
@@ -127,21 +124,22 @@ export interface RecipientInfo {
...
+): Recipient {
   const e164PhoneNumber = info.addressToE164Number[address]
   const numberRecipient = e164PhoneNumber ? info.phoneRecipientCache[e164PhoneNumber] : undefined
   const valoraRecipient = info.valoraRecipientCache[address]
-  const displayInfo = info.addressToDisplayName[address]
 
   const recipient: Recipient = {
     address,
-    name: valoraRecipient?.name || numberRecipient?.name || displayInfo?.name,
-    thumbnailPath: valoraRecipient?.thumbnailPath || displayInfo?.imageUrl || undefined,
```
Comment (jeanregisser) [inline packages/mobile/src/recipients/recipient.ts:144]: We need to keep `addressToDisplayName` for correctly showing Bidali transactions.

This makes me think we likely want a migration to only keep Bidali addresses [in the identify reducer](https://github.com/valora-inc/wallet/blob/cdf364a68b3745de20b7357dcf3ce8b14449fbe4/packages/mobile/src/identity/reducer.ts#L100).

File: packages/mobile/src/firebase/notifications.ts [hunk 2]
```diff
@@ -72,6 +72,7 @@ function* handlePaymentReceived(
         },
         address: transferNotification.sender.toLowerCase(),
         comment: transferNotification.comment,
+        // TODO: Add server default value from backend notification service
```
Comment (jeanregisser) [inline packages/mobile/src/firebase/notifications.ts:75]: 👍

File: packages/mobile/src/transactions/CeloTransferFeedItem.test.tsx [hunk 3]
```diff
@@ -75,6 +77,31 @@ describe('CeloTransferFeedItem', () => {
...
+          defaultImage={null}
+          defaultName={null}
+          {...getMockI18nProps()}
+        />
+      </Provider>
+    )
+    expect(tree).toMatchSnapshot()
+  })
+
+  it("renders correctly when there's a default name", () => {
```
Comment (jeanregisser) [inline packages/mobile/src/transactions/CeloTransferFeedItem.test.tsx:89]: Nothing to do now, but wanted to point out that tests that only check the render tree via a snapshot comparison are often not very useful.
We're trying to move toward more meaningful checks via `react-native-testing-libray`.
Specifically checking for the presence of text, buttons, etc.
Example: https://github.com/valora-inc/wallet/blob/69daaa7ba298ea7daa04cd8a976c749d572e8e0f/packages/mobile/src/backup/DelayButton.test.tsx#L29

File: packages/mobile/src/transactions/TransferConfirmationCard.tsx [hunk 4]
```diff
@@ -34,19 +34,19 @@ import UserSection from 'src/transactions/UserSection'
 import { Currency } from 'src/utils/currencies'
 import { navigateToURI } from 'src/utils/linking'
 
+// DIEGO: I think we can remove address and phone number from here, since we have that info in recipient.
```
Comment (jeanregisser) [inline packages/mobile/src/transactions/TransferConfirmationCard.tsx:37]: Yes it seems so 👍

File: packages/mobile/src/transactions/TransferConfirmationCard.tsx [hunk 5]
```diff
@@ -88,8 +88,8 @@ function InviteSentContent({ addressHasChanged, recipient, amount }: Props) {
       <UserSection
         type="sent"
         addressHasChanged={addressHasChanged}
-        recipient={recipient}
-        avatar={<ContactCircle recipient={recipient} />}
+        recipient={recipient!}
+        avatar={<ContactCircle recipient={recipient!} />}
```
Comment (jeanregisser) [inline packages/mobile/src/transactions/TransferConfirmationCard.tsx:92]: Can we avoid `!`?

File: packages/mobile/src/transactions/transferFeedUtils.ts [hunk 6]
```diff
@@ -224,7 +231,7 @@ export function getTransferFeedParams(
   return { title, info, recipient }
 }
 
-export function getTxsFromUserTxQuery(data: UserTransactionsQuery | undefined) {
+export function getTxsFromUserTxQuery(data?: UserTransactionsQuery) {
```
Comment (jeanregisser) [inline packages/mobile/src/transactions/transferFeedUtils.ts:234]: Why this change?
Will be functionally the same at runtime, but semantically it's different, and will change the TypeScript check.
The previous one required a parameter to be passed (even if undefined), the changed version makes the parameter optional.
In this case, expecting data to be passed seems what we want.

File: packages/mobile/src/transactions/TransferConfirmationCard.tsx [hunk 7]
```diff
@@ -220,8 +220,8 @@ function RewardContent({ amount, recipient }: Props) {
       <UserSection
         type="received"
         expandable={false}
-        recipient={recipient}
-        avatar={<TransferAvatars type="received" recipient={recipient} />}
+        recipient={recipient!}
+        avatar={<TransferAvatars type="received" recipient={recipient!} />}
```
Comment (jeanregisser) [inline packages/mobile/src/transactions/TransferConfirmationCard.tsx:224]: Same as above about `!`

**Review by jeanregisser — COMMENTED** [review]
Great work! 👍 

Could you stay closer to our [original PR template](https://github.com/valora-inc/wallet/blob/main/.github/pull_request_template.md) and link to the issue in the PR description instead?

File: packages/blockchain-api/src/helpers/KnownAddressesCache.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+import { database } from '../firebase'
+
+const TAG = 'helpers/KnownAddressesCache'
+
+const ROOT_KEY = 'addressesExtraInfo'
+
+const ON_VALUE_CHANGED = 'value'
+
+class KnownAddressesCache {
+  private knownAddresses: { [address: string]: { name?: string; imageUrl?: string } } = {}
```
[the remark this answers is not in the document]
Comment (dievazqu) [reply packages/blockchain-api/src/helpers/KnownAddressesCache.ts:11]: I was returning `{}` when we don't have any info, so we can unpack like:
```
{name, imageUrl} = getValueFor(address)
```
And get undefined values for `name` and `imageUrl` instead of an error if there is no info available.
I can change that and return `undefined` but then I would need to check if the returning value is valid before unpacking. What do you think is the best approach?

File: packages/mobile/src/recipients/recipient.ts (hunk 1 above)
Comment (jeanregisser) [reply packages/mobile/src/recipients/recipient.ts]: Oh yeah good idea!

File: packages/blockchain-api/src/helpers/KnownAddressesCache.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeanregisser) [reply packages/blockchain-api/src/helpers/KnownAddressesCache.ts:11]: Yes what I meant is without the ` | undefined` the lookup could seems it's working, while it would crash at runtime.
Basically makes sure we check the result where used (which you already did, but not because TypeScript warned you).
See [this playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBA1mEB3MBBAJugTgUwhPALhgG8YBtAQ0132OiwEswBzAXWLLEoFscB+elCasA3DEY9KLHAFUsAG0EwGzFjAC+mmAF5SGgLAAoYwDMArmGBRG4GDKgA1SgvM4AYiCwAKatjwQQiIsAJScMNx8yqpiElIy8kpBatokxjAwoJCwAG4ubrrwiCgY-vh4VDQBbMbpMLhQ5lhgMHmuOMYaQA) and [this one](https://www.typescriptlang.org/play?#code/MYewdgzgLgBA1mEB3MBBAJugTgUwhPALhgG8YBtAQ0132OiwEswBzAXWLLEoFscB+elCasA3DEY9KLHAFUsAG0EwGzFjAC+MAD4wArmHQ4AZsxzpNMALykNAWABQj4weBRG4GDKgA1Sgr0cADEQLAAKamw8CCERFgBKThhuPmVVMQkpGXklWLVLEkcYGFBIWAA3f0DreEQUDCj8PCoaaLZHIphcKD0sMBhKgJxHDSA) for the difference.

[the remark this answers is not in the document]
Comment (dievazqu) [reply packages/blockchain-api/src/helpers/KnownAddressesCache.ts:11]: Ohhh, got it! Thanks!

**Review by jeanregisser — APPROVED** [review]
🎉

**Review by gnardini — COMMENTED** [review]
Great work! Left a few minor comments.

Sorry for taking so long to review, I will be better abut that in the future!

File: packages/mobile/src/recipients/recipient.ts [hunk 9]
```diff
@@ -132,16 +133,27 @@ export interface RecipientInfo {
   addressToDisplayName: AddressToDisplayNameType
 }
 
-export function getRecipientFromAddress(address: string, info: RecipientInfo) {
+export function getRecipientFromAddress(
+  address: string,
+  info: RecipientInfo,
+  defaultName?: Maybe<string>,
```
[the remark this answers is not in the document]
Comment (dievazqu) [reply packages/mobile/src/recipients/recipient.ts:139]: Yep! You are right, I was a bit confused between `null` and `undefined` values here.
But I'm gonna double check this, actually defaultName and defaultImage should be nullable :)

**Review by gnardini — APPROVED** [review]
🚀
